File: WebsiteHoTroDichVuBanDoc_Backend/WebsiteHoTroDichVuBanDoc_Backend/app/models/muon_tra.py
from typing import Optional
from datetime import datetime, date
from decimal import Decimal
from app.models.db_base import DBModel

class MuonTraBase(DBModel):
    maBanSao: int
    maBanDoc: int
    maNhanVien: Optional[int] = None # Nhân viên thực hiện cho mượn
    ngayTra: date # Ngày hẹn trả
    trangThaiMuon: str = 'daMuon'
    soLanGiaHan: int = 0
    soLanGiaHanToiDa: int = 2
    tienPhat: Optional[Decimal] = 0
    ghiChu: Optional[str] = None

class MuonTraCreate(MuonTraBase):
    # Khi tạo mới, chỉ cần các thông tin cơ bản
    # ngayTraThucTe sẽ là None
    pass

class MuonTraUpdate(DBModel):
    # Dùng khi trả sách hoặc cập nhật
    maNhanVien: Optional[int] = None # Nhân viên nhận sách trả
    ngayTraThucTe: Optional[datetime] = None
    trangThaiMuon: Optional[str] = None
    tienPhat: Optional[Decimal] = None
    ghiChu: Optional[str] = None

class MuonTra(MuonTraBase):
    maMuonTra: int
    thoiGianMuon: Optional[datetime] = None
    ngayTraThucTe: Optional[datetime] = None

class MuonTraTraSach(DBModel):
    maNhanVien: int # Nhân viên nhận sách trả

class MuonTraItem(DBModel):
    maMuonTra: int
    tenTacPham: str
    maBanSaoNoiBo: str
    ngayMuon: datetime # Timestamptz
    ngayTraDuKien: date
    ngayTraThucTe: Optional[datetime] = None
    trangThai: str # daMuon, daTra...
    tienPhat: Optional[float] = 0.0
    nguoiMuon: Optional[str] = None # Trường này để Nhân viên biết ai mượn

File: WebsiteHoTroDichVuBanDoc/Backend/src/main/java/com/library/website/models/MuonTra.java
package com.library.website.models;

import jakarta.persistence.*;
import lombok.Getter;
import lombok.Setter;
import lombok.NoArgsConstructor;
import lombok.AllArgsConstructor;
import java.math.BigDecimal;
import java.time.LocalDate;
import java.time.LocalDateTime;

@Entity
@Table(name = "muontra", schema = "public")
@Getter
@Setter
@NoArgsConstructor
@AllArgsConstructor
public class MuonTra {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Column(name = "mamuontra")
    private Long maMuonTra;

    @ManyToOne
    @JoinColumn(name = "mabansao")
    private BanSao banSao;

    @ManyToOne
    @JoinColumn(name = "mabandoc")
    private BanDoc banDoc;

    @Column(name = "manhanvien")
    private Long maNhanVien;

    @Column(name = "thoigianmuon")
    private LocalDateTime thoiGianMuon = LocalDateTime.now();

    @Column(name = "ngaytra")
    private LocalDate ngayTra;

    @Column(name = "ngaytrathucte")
    private LocalDateTime ngayTraThucTe;

    @Column(name = "trangthaimuon")
    private String trangThaiMuon = "daMuon"; // 'daMuon', 'daTra', 'quaHan', 'dangChoXacNhan'

    @Column(name = "solangiahan")
    private Short soLanGiaHan = 0;

    @Column(name = "solangiahantoida")
    private Short soLanGiaHanToiDa = 2;

    @Column(name = "tienphat")
    private BigDecimal tienPhat = BigDecimal.ZERO;

    @Column(name = "ghichu")
    private String ghiChu;

    @Column(name = "updated_at")
    private LocalDateTime updatedAt = LocalDateTime.now();

    @PreUpdate
    public void preUpdate() {
        this.updatedAt = LocalDateTime.now();
    }
}
